### DISCUSSION TITLE
apply an effect (ADSRGain) over another effect (DELAY) over a Kit stream

### DISCUSSION THREAD

**Discussion by PantalaLabs** [pschatzmann/arduino-audio-tools#397] [Q&A] [answered]
Hi folks.

This is my first post.
I´d like to thank you Phil Schatzmann for this amazing work , and all you guys for the effort to keep threads solved.

My question is about applying a ADSRGain over a Delay.
I can create and set the initial objects for the delays and adsrs , yes I have some of them.

I´ve searched examples and found this
[https://github.com/pschatzmann/arduino-audio-tools/wiki/Audio-Effects](https://github.com/pschatzmann/arduino-audio-tools/wiki/Audio-Effects)
And found a similar ADSRgain discussion , but with one layer.

I´m struggling to apply an effect (ADSRGain) over another effect (DELAY) that is over a Kit stream

```

#include <AudioTools.h>
#include <AudioLibs/AudioKit.h>
#include <AudioLibs/MemoryManager.h>
#define MAXGRAINS 5


uint16_t sample_rate = 22050;
uint8_t channels = 1;
AudioKitStream kit;
MemoryManager memory(500);           // Activate SPI RAM for objects > 500 bytes
DynamicMemoryStream recording(true); // Audio stored on heap looping endlessly
AudioEffects<GeneratorFromStream<effect_t>> effects(kit, 2, 1.0);
GeneratedSoundStream<int16_t> in(effects);

Delay delay0, delay1, delay2, delay3, delay4;
Delay *delays[MAXGRAINS] = {&delay0, &delay1, &delay2, &delay3, &delay4};

ADSRGain adsr0, adsr1, adsr2, adsr3, adsr4;
ADSRGain *adsrs[MAXGRAINS] = {&adsr0, &adsr1, &adsr2, &adsr3, &adsr4};

StreamCopy copier(kit, in); // copy in to out


void setup(void)
{
  Serial.begin(115200);

  for (uint8_t i = 0; i < MAXGRAINS; i++)
  {
    effects.addEffect(delays[i]);
    delays[i]->setDuration(random(200, 350));
    delays[i]->setDepth(random(85, 95) / 100.0);
    delays[i]->setFeedback(random(75, 85) / 100.0);
    delays[i]->setSampleRate(sample_rate);

    effects.addEffect(adsrs[i]);
    adsrs[i]->setAttackRate(0.001);
    adsrs[i]->setDecayRate(0.0001);
    adsrs[i]->setSustainLevel(0.8);
    adsrs[i]->setReleaseRate(0.005);
    
  }

  AudioLogger::instance().begin(Serial, AudioLogger::Error);

  auto cfg = kit.defaultConfig(RXTX_MODE);
  cfg.sample_rate = sample_rate;
  cfg.channels = channels;
  cfg.input_device = AUDIO_HAL_ADC_INPUT_LINE2;
  cfg.bits_per_sample = sizeof(int16_t) * 8;
  cfg.sd_active = false;
  kit.begin(cfg);
  kit.setVolume(0.7);
}



```

Any help would be appreciated.

Thank you 
Gibran

**Comment by pschatzmann** [comment, marked as answer]
I don't understand why you add so many offects of the same type in an array. 
Usually you want to use just one effect of a kind and adjust the corresponding parameters

**Reply by PantalaLabs** [reply]
Hi Phil 

Thank you for your reply.
This is a [granular processor](https://en.wikipedia.org/wiki/Granular_synthesis).
Basically , it takes many tiny parts - grains- of the incoming sound and repeat them over the time.
I already did some experiments in the past with the  Arduino Due + coding , but , the result wasn´t so good due to audio quality.
Last month I discovered the ESP Audio Kit board, and last week , your library . 
So , I think that my 3 year old dream is about to come true.

The grains are easy to control like :
```

for (uint8_t i = 0; i < MAXGRAINS; i++)
  {
    if (millis() > activate[i])
    {
      activate[i] = millis() + MINGRAININTERVAL + random(5, 100);
      delays[i]->setActive(!delays[i]->active());
    }
  }

```
Last night they worked like a charm. 
I can dinamically control  Duration, Interval, Depth and Fedback of each grain.
The random and all parameters must be replaced by an user variable control like an analog reading (I know that its not audio kit built in , but I´ll think about it later , maybe MIDI BLE or something)
Each grain must have its own calculated adsr (it will be calculated with activate[i] size)

For now , I cant solve the need to patch each grain through its envelope to make it sound smoother , by firing the adsr (triangle shaped) every time the grain is set to active.

Thank you 

Gibran

**Comment by pschatzmann** [comment]
Cool, I was not aware of this !

**Reply by PantalaLabs** [reply]
This weekend I'll try to code it
As soon as I make it work , I'll update the thread .

Thank you !!

**Comment by pschatzmann** [comment]
You could also provide a web dialog for updating the parameters:
I created [this example](https://github.com/pschatzmann/TinyHttp/blob/main/examples/webservice/webservice-html/webservice-html.ino) as a proof of concept...

**Comment by PantalaLabs** [comment]
Nice solution ! Thanks!

I dont know if you´ve seen it,  there is a mod to replace the switch buttons by potentiometers to make analogreads.

Today I´ve been working into serial communication , read analog values with a Nano and send to ESP32 Kit via TX/RX but , the RX pin of my board keeps always HI.

Any advance, I´ll share.

Thanks !!
